### PR TITLE
Makes the preloader apply text macros: Re-engineering

### LINKED ISSUE #24615: Map loader does not respect text macros in var edits
Revealed in #24171 
> \improper Garden APC

### PR DESCRIPTION
**Pull request by Cyberboss**
Fixes #24615
Closes #24619

@RemieRichards 

:cl: Cyberboss
fix: You should no longer be seeing entities with `\improper` in front of their name
/:cl:

### CODE REVIEW

**Comment by RemieRichards** [comment]
You're only doing one macro per string?

File: code/__HELPERS/text.dm [hunk 1]
```diff
@@ -554,3 +554,57 @@ var/list/binary = list("0","1")
...
+	var/base = copytext(string, 1, next_backslash)
+	var/macro = lowertext(copytext(string, next_backslash + 1, next_space))
+	var/rest = copytext(string, next_space)
+
+
+	//See http://www.byond.com/docs/ref/info.html#/DM/text/macros
+	switch(macro)
+		//prefixes/agnostic
+		if("the")
+			rest = text("\the []", rest)
```
Comment (RemieRichards) [inline code/__HELPERS/text.dm]: Oh how I wish we could just do this:
```
var/static/list/prefix_macros = list("the" = TRUE, "a" = TRUE, "an" = TRUE ...)
var/static/list/suffix_macros = list("s" = TRUE, "he" = TRUE, "his" = TRUE ...)
if(prefix_macros[macro])
    rest = text("\[] []", macro, rest)
else if(suffix_macros[macro])
    base = text("[]\[]", rest, macro)
```

**Comment by AnturK** [comment]
Might be worth opening a feature request for inbuilt macro reeval on byond forums.
But for now this will have to do.

**Comment by Cyberboss** [comment]
Closing in favor of new PR

**Comment by RemieRichards** [comment]
@MrStonedOne that's deprecated, I don't think we should rely on it.

**Comment by MrStonedOne** [comment]
it's the least chinsky way to do that

**Comment by AnturK** [comment]
@tgstation/commit-access Opinions please.

**Comment by optimumtact** [comment]
I don't think we review map changes nearly enough to trust an eval on map objects

**Comment by optimumtact** [comment]
I should read pr's before I comment

**Comment by MrStonedOne** [comment]
eval should only work on expressions, like 1+1 or "\ref[thing]" I don't think that it can actually be used for an exploit.

**Comment by RemieRichards** [comment]
that's a challenge, where's our resident footwear?

**Comment by optimumtact** [comment]
the eval is gone guys

**Comment by Cyberboss** [comment]
@optimumtact there's 2 opposite prs

On Mar 6, 2017 5:07 PM, "oranges" <notifications@github.com> wrote:

the eval is gone guys

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/24614#issuecomment-284549276>,
or mute the thread

**Comment by MrStonedOne** [comment]
I'm arguing that we do eval() because making a new version of byond's code sucks and because I get the feeling eval will be useful some time in the future so some testing now would be neat.

**Pull request merged**